openmc/model/triso.py

```diff
@@ -153,3 +520,502 @@ def create_triso_lattice(trisos, lower_left, pitch, shape, background):
+        rods so rods can be "removed" without breaking the heap structure
+        invariant.
+
+        """
+
+        # Create KD tree for quick nearest neighbor search
+        tree = scipy.spatial.cKDTree(particles)
+
+        # Find distance to nearest neighbor and index of nearest neighbor for
+        # all particles
+        d, n = tree.query(particles, k=2)
+        d = d[:,1]
+        n = n[:,1]
+
+        # Array of particle indices, indices of nearest neighbors, and
+        # distances to nearest neighbors
+        a = np.dstack(([i for i in range(len(n))], n, d))[0]
+
+        # Array of nearest neighbor indices, indices of particles they are
+        # nearest neighbors of, and distances between them
+        b = a[a[:,1].argsort()]
+        b[:,[0, 1]] = b[:,[1, 0]]
+
+        # Find the intersection between 'a' and 'b': a list of particles who
+        # are each other's nearest neighbors and the distance between them
+        r = [x for x in {tuple(x) for x in a} & {tuple(x) for x in b}]
+
+        # Remove duplicate rods and sort by distance
+        r = map(list, set([(x[2], int(min(x[0:2])), int(max(x[0:2])))
+                            for x in r]))
+
+        # Clear priority queue and add rods
+        del rods[:]
+        rods_map.clear()
+        for d, i, j in r:
+            add_rod(d, i, j)
+
+        # Inner diameter is set initially to the shortest center-to-center
+        # distance between any two particles
+        if rods:
+            inner_diameter[0] = rods[0][0]
+
+
+    def update_mesh(i):
+        """Update which mesh cells the particle is in based on new particle
+        center coordinates.
+
+        'mesh'/'mesh_map' is a two way dictionary used to look up which
+        particles are located within one diameter of a given mesh cell and
+        which mesh cells a given particle center is within one diameter of.
+        This is used to speed up the nearest neighbor search.
+
+        Parameters
+        ----------
+        i : int
+            Index of particle in particles array.
+
+        """
+
+        # Determine which mesh cells the particle is in and remove the
+        # particle id from those cells
+        for idx in mesh_map[i]:
+            mesh[idx].remove(i)
+        del mesh_map[i]
+
+        # Determine which mesh cells are within one diameter of particle's
+        # center and add this particle to the list of particles in those cells
+        for idx in domain.nearby_mesh_cells(particles[i]):
+            mesh[idx].add(i)
+            mesh_map[i].add(idx)
+
+
+    def reduce_outer_diameter():
+        """Reduce the outer diameter so that at the (i+1)-st iteration it is:
+
+            d_out^(i+1) = d_out^(i) - (1/2)^(j) * d_out0 * k / n,
+
+        where k is the contraction rate, n is the number of particles, and
+
+            j = floor(-log10(pf_out - pf_in)).
+
+        """
+
+	inner_pf = (4/3 * pi * (inner_diameter[0]/2)**3 * n_particles /
```

Comment:
You have a tab character at the beginning of this line (actually an error for Python 3!)